File: examples/composite-exo-glove/python/src/telemetry.py
from __future__ import annotations

import argparse

from qnbot_sdk import (
    CompositeExoGloveConfig,
    DeviceStatus,
    Sample,
    Sdk,
    SerialConnection,
)
from qnbot_sdk.exo import ExoConfig, ExoTelemetry
from qnbot_sdk.glove import GloveConfig, GlovePose

EXO_NAME = "exo"
GLOVE_NAME = "glove"


def create_sdk(port: str) -> Sdk:
    return Sdk(
        devices=(
            CompositeExoGloveConfig(
                connection=SerialConnection(port=port),
                devices=(GloveConfig(name=GLOVE_NAME), ExoConfig(name=EXO_NAME)),
            ),
        )
    )


def print_glove(sample: Sample[GlovePose]) -> None:
    print(
        f"glove pose sequence={sample.sequence} hand={sample.value.payload.hand_side}"
    )


def print_exo(sample: Sample[ExoTelemetry]) -> None:
    print(f"exo telemetry sequence={sample.sequence} quality={sample.value.quality}")


def print_status(sample: Sample[DeviceStatus]) -> None:
    print(f"exo status sequence={sample.sequence} value={sample.value}")


def print_glove_status(sample: Sample[DeviceStatus]) -> None:
    print(f"glove status sequence={sample.sequence} value={sample.value}")


def main() -> None:
    arguments = argparse.ArgumentParser(
        description="Read combined telemetry and status"
    )
    arguments.add_argument("--port", required=True, help="Shared serial port")
    options = arguments.parse_args()

    sdk = create_sdk(port=options.port)
    try:
        glove_domain = sdk.glove()
        exo_domain = sdk.exo()
        glove = glove_domain.device(GLOVE_NAME)
        exo = exo_domain.device(EXO_NAME)
        glove.pose().subscribe(print_glove)
        glove.status().subscribe(print_glove_status)
        exo.telemetry().subscribe(print_exo)
        exo.status().subscribe(print_status)
        glove_domain.start()
        exo_domain.start()
        print("running; press Ctrl+C to stop", flush=True)
        # The aggregate runner is required to drive both member domains.
        sdk.run_forever()
    except KeyboardInterrupt:
        glove_domain.stop()
        exo_domain.stop()
    finally:
        glove_domain.close()
        exo_domain.close()


if __name__ == "__main__":
    main()

File: composite-exo-glove/cpp/src/telemetry.cpp
#include "composite_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>

int main(int argc, char** argv) {
    try {
        qnbot::Sdk sdk(
            example::composite_config(example::parse_port(argc, argv)));
        auto glove_domain = sdk.glove();
        auto exo_domain = sdk.exo();
        auto glove = glove_domain.device("glove");
        auto exo = exo_domain.device("exo");
        const auto pose_subscription = glove.pose().subscribe(
            [](const qnbot::Sample<qnbot::GlovePose>& sample) {
                std::cout << "glove pose sequence=" << sample.sequence
                          << " fingers="
                          << sample.value.payload.fingertip_local.size()
                          << '\n';
            });
        const auto glove_status_subscription = glove.status().subscribe(
            [](const qnbot::Sample<qnbot::GloveStatus>& sample) {
                std::cout << "glove status connected=" << sample.value.connected
                          << " stale=" << sample.value.stale << '\n';
            });
        const auto exo_subscription = exo.telemetry().subscribe(
            [](const qnbot::Sample<qnbot::ExoTelemetry>& sample) {
                std::cout << "exo telemetry sequence=" << sample.sequence
                          << " left_arm=";
                example::print_values(
                    sample.value.payload.left_arm.joint_positions_rad);
                std::cout << '\n';
            });
        const auto exo_status_subscription = exo.status().subscribe(
            [](const qnbot::Sample<qnbot::DeviceStatus>& sample) {
                std::cout << "exo status connected=" << sample.value.connected
                          << " stale=" << sample.value.stale << '\n';
            });

        glove_domain.start();
        exo_domain.start();
        std::cout << "reading Glove pose, Exo telemetry, and both statuses; "
                     "press Ctrl+C to stop\n";
        // Drive both member domains through the aggregate runner.
        sdk.run_forever();
        static_cast<void>(pose_subscription);
        static_cast<void>(glove_status_subscription);
        static_cast<void>(exo_subscription);
        static_cast<void>(exo_status_subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
